File: cli/ruby/agent_loader.rb
# Agent framework loader for Ruby CLI tools.
#
# This module provides utilities to load and use agents, tools, roles, rules,
# and templates from the main project structure.

require 'yaml'
require 'pathname'

class AgentLoader
  def initialize
    @cli_path = File.expand_path('../', __FILE__)
  end

  # Load an agent configuration by name.
  #
  # @param agent_name [String] Name of the agent (e.g., 'coder-agent')
  # @return [Hash, nil] Agent configuration or nil if not found
  def load_agent(agent_name)
    role_path = File.join(@cli_path, 'role', agent_name, 'role.yaml')

    unless File.exist?(role_path)
      puts "Error: Agent '#{agent_name}' not found at #{role_path}"
      return nil
    end

    parse_yaml(role_path)
  end

  # Load a tool configuration by name.
  #
  # @param tool_name [String] Name of the tool (e.g., 'write_python')
  # @return [Hash, nil] Tool configuration or nil if not found
  def load_tool(tool_name)
    tool_path = File.join(@cli_path, 'tool', "#{tool_name}.yaml")

    if File.exist?(tool_path)
      return parse_yaml(tool_path)
    end

    # Search recursively in tool subdirectories
    tool_dir = File.join(@cli_path, 'tool')
    files = find_yaml_files(tool_dir, tool_name)

    unless files.empty?
      return parse_yaml(files[0])
    end

    puts "Error: Tool '#{tool_name}' not found"
    nil
  end

  # List all available agents.
  #
  # @return [Array<String>] List of agent names
  def list_agents
    role_dir = File.join(@cli_path, 'role')
    agents = []

    if Dir.exist?(role_dir)
      Dir.each_child(role_dir) do |item|
        item_path = File.join(role_dir, item)
        if File.directory?(item_path) && File.exist?(File.join(item_path, 'role.yaml'))
          agents << item
        end
      end
    end

    agents.sort
  end

  # List all available tools.
  #
  # @return [Array<String>] List of tool names
  def list_tools
    tool_dir = File.join(@cli_path, 'tool')
    tools = []

    if Dir.exist?(tool_dir)
      find_all_yaml_tools(tool_dir, tools)
    end

    tools.sort.uniq
  end

  private

  # Parse a simple YAML file.
  #
  # @param file_path [String] Path to YAML file
  # @return [Hash] Parsed YAML as hash
  def parse_yaml(file_path)
    YAML.load_file(file_path) || {}
  rescue => e
    puts "Error parsing YAML at #{file_path}: #{e.message}"
    {}
  end

  # Find YAML files recursively.
  #
  # @param dir [String] Directory to search
  # @param tool_name [String] Tool name to find
  # @return [Array<String>] List of matching file paths
  def find_yaml_files(dir, tool_name)
    files = []

    return files unless Dir.exist?(dir)

    Dir.each_child(dir) do |item|
      path = File.join(dir, item)

      if File.directory?(path)
        files.concat(find_yaml_files(path, tool_name))
      elsif File.file?(path) && File.extname(path) == '.yaml'
        if File.basename(path, '.yaml') == tool_name
          files << path
        end
      end
    end

    files
  end

  # Find all YAML tool files.
  #
  # @param dir [String] Directory to search
  # @param tools [Array] Reference to tools array
  def find_all_yaml_tools(dir, tools)
    return unless Dir.exist?(dir)

    Dir.each_child(dir) do |item|
      path = File.join(dir, item)

      if File.directory?(path)
        find_all_yaml_tools(path, tools)
      elsif File.file?(path) && File.extname(path) == '.yaml'
        tools << File.basename(path, '.yaml')
      end
    end
  end
end

File: cli/php/AgentLoader.php
<?php

/**
 * Agent framework loader for PHP CLI tools.
 *
 * This class provides utilities to load and use agents, tools, roles, rules,
 * and templates from the main project structure via symlinks.
 */

class AgentLoader {
    private $cliPath;
    
    public function __construct() {
        $this->cliPath = dirname(__FILE__) . '/..';
    }
    
    /**
     * Load an agent configuration by name.
     *
     * @param string $agentName Name of the agent (e.g., 'coder-agent')
     * @return array|null Agent configuration or null if not found
     */
    public function loadAgent($agentName) {
        $rolePath = $this->cliPath . '/role/' . $agentName . '/role.yaml';

        if (!file_exists($rolePath)) {
            echo "Error: Agent '{$agentName}' not found at {$rolePath}\n";
            return null;
        }

        return $this->parseYaml($rolePath);
    }
    
    /**
     * Load a tool configuration by name.
     *
     * @param string $toolName Name of the tool (e.g., 'write_python')
     * @return array|null Tool configuration or null if not found
     */
    public function loadTool($toolName) {
        $toolPath = $this->cliPath . '/tool/' . $toolName . '.yaml';

        if (file_exists($toolPath)) {
            return $this->parseYaml($toolPath);
        }

        // Search recursively in tool subdirectories
        $toolDir = $this->cliPath . '/tool';
        $files = $this->findYamlFiles($toolDir, $toolName);

        if (!empty($files)) {
            return $this->parseYaml($files[0]);
        }

        echo "Error: Tool '{$toolName}' not found\n";
        return null;
    }
    
    /**
     * List all available agents.
     *
     * @return array List of agent names
     */
    public function listAgents() {
        $roleDir = $this->cliPath . '/role';
        $agents = [];

        if (is_dir($roleDir)) {
            foreach (scandir($roleDir) as $item) {
                if ($item !== '.' && $item !== '..' && is_dir($roleDir . '/' . $item)) {
                    if (file_exists($roleDir . '/' . $item . '/role.yaml')) {
                        $agents[] = $item;
                    }
                }
            }
        }

        sort($agents);
        return $agents;
    }
    
    /**
     * List all available tools.
     *
     * @return array List of tool names
     */
    public function listTools() {
        $toolDir = $this->cliPath . '/tool';
        $tools = [];

        if (is_dir($toolDir)) {
            $this->findAllYamlTools($toolDir, $tools);
        }

        sort($tools);
        return array_unique($tools);
    }
    
    /**
     * Parse a simple YAML file.
     * 
     * @param string $filePath Path to YAML file
     * @return array Parsed YAML as associative array
     */
    private function parseYaml($filePath) {
        $content = file_get_contents($filePath);
        $lines = explode("\n", $content);
        $result = [];
        
        foreach ($lines as $line) {
            $line = trim($line);
            if (empty($line) || strpos($line, '#') === 0) {
                continue;
            }
            
            if (strpos($line, ':') !== false) {
                list($key, $value) = explode(':', $line, 2);
                $result[trim($key)] = trim($value);
            }
        }
        
        return $result;
    }
    
    /**
     * Find YAML files recursively.
     * 
     * @param string $dir Directory to search
     * @param string $toolName Tool name to find
     * @return array List of matching file paths
     */
    private function findYamlFiles($dir, $toolName) {
        $files = [];
        
        if (!is_dir($dir)) {
            return $files;
        }
        
        foreach (scandir($dir) as $item) {
            if ($item === '.' || $item === '..') {
                continue;
            }
            
            $path = $dir . '/' . $item;
            
            if (is_dir($path)) {
                $files = array_merge($files, $this->findYamlFiles($path, $toolName));
            } elseif (is_file($path) && pathinfo($path, PATHINFO_EXTENSION) === 'yaml') {
                if (pathinfo($path, PATHINFO_FILENAME) === $toolName . '.yaml') {
                    $files[] = $path;
                }
            }
        }
        
        return $files;
    }
    
    /**
     * Find all YAML tool files.
     * 
     * @param string $dir Directory to search
     * @param array &$tools Reference to tools array
     */
    private function findAllYamlTools($dir, &$tools) {
        if (!is_dir($dir)) {
            return;
        }
        
        foreach (scandir($dir) as $item) {
            if ($item === '.' || $item === '..') {
                continue;
            }
            
            $path = $dir . '/' . $item;
            
            if (is_dir($path)) {
                $this->findAllYamlTools($path, $tools);
            } elseif (is_file($path) && pathinfo($path, PATHINFO_EXTENSION) === 'yaml') {
                $tools[] = pathinfo($path, PATHINFO_FILENAME);
            }
        }
    }
}

?>
